### PR TITLE
Extract workspace-base from workspace-full

### PR DESCRIPTION
**Pull request by aledbf**
The changes in the pull request also reduce the size of the images (between ~200MB and ~600MB) 

```console

gitpod/workspace-full                latest               3db28764b78e   13 days ago          6.48GB
gitpod/workspace-full                latest               21ddaeec4cd5   9 minutes ago        5.88GB

gitpod/workspace-full-vnc            latest               711deb23bbcf   13 days ago          6.93GB
gitpod/workspace-full-vnc            latest               113c3f8d86f6   16 seconds ago       6.63GB

gitpod/workspace-dotnet              latest               985b4054e5dc   13 days ago          6.87GB
gitpod/workspace-dotnet              latest               ae92e9265595   7 seconds ago        6.27GB

gitpod/workspace-dotnet-vnc          latest               2897c0721492   13 days ago          7.32GB
gitpod/workspace-dotnet-vnc          latest               113c3f8d86f6   4 minutes ago        6.63GB

gitpod/workspace-gecko               latest               a574d8af3ec3   13 days ago          8.79GB
gitpod/workspace-gecko               latest               30b2b1653654   6 seconds ago        8.04GB

gitpod/mdbook                        latest               99a2b1d122f3   13 days ago          6.57GB
gitpod/mdbook                        latest               e7bedc55e5fb   12 seconds ago       5.92GB

gitpod/workspace-mongodb             latest               31407ff460d7   13 days ago          6.68GB
gitpod/workspace-mongodb             latest               090acf64f6ff   7 seconds ago        6.08GB

gitpod/workspace-mysql               latest               317ce94af1be   13 days ago          7.01GB
gitpod/workspace-mysql               latest               8e0e7db50de7   2 minutes ago        6.26GB

gitpod/workspace-postgres            latest               349c366bc467   13 days ago          6.64GB
gitpod/workspace-postgres            latest               bad8de157f55   17 seconds ago       6.04GB

gitpod/workspace-wasm                latest               00f9c0138b6c   13 days ago          8.39GB
gitpod/workspace-wasm                latest               e7fdf6d9e0d2   30 seconds ago       7.88GB

```

replaces #272
fixes https://github.com/gitpod-io/gitpod/issues/2694
fixes #263

### CODE REVIEW

File: flutter/Dockerfile
Comment (jankeromnes) [inline flutter/Dockerfile:15]: Ah, awesome, thanks for this drive-by fix! 💯

File: base/install-packages
Comment (jankeromnes) [inline base/install-packages:20]: Adding `--no-install-recommends` will save some space, but also means that many packages that were previously installed as a side effect will now be missing.

This might inadvertently break downstream images / project setups that might have relied on these packages to be present.

While we could arguably consider this an "end-user problem", I think we should extensively test the images produced by this change, for example against the 5 most common projects on Gitpod.

File: full/Dockerfile
Comment (jankeromnes) [inline full/Dockerfile:10]: Side note: Not really important, but incrementing this is unnecessary if you already touch other lines in the same "Dazzle layer" (e.g. here you're touching `apt-get update` below, which will already invalidate the cache).

File: full/Dockerfile
Comment (jankeromnes) [inline full/Dockerfile:57]: 👍 I like this change (simpler / less awkward is usually better)

File: full/Dockerfile
Comment (jankeromnes) [inline full/Dockerfile:71]: Whoops. This line actually intended to _remove_ the `cmake` installed by `apt-get` (so that it can be replaced by the one installed with `brew`).

But maybe this isn't even needed anymore? We should check which version is installed by `apt-get` and which by `brew`. If `apt-get` already gives the latest stable, we can remove these two lines.

File: full/Dockerfile
Comment (jankeromnes) [inline full/Dockerfile:85]: Please rebase on top of latest `master` (this change should automatically disappear)

File: full/Dockerfile
Comment (jankeromnes) [inline full/Dockerfile:214]: Oops. Changing any lines in the "Docker Dazzle layer" will run into the same problem as https://github.com/gitpod-io/workspace-images/pull/331 i.e. it will upgrade the version of `docker` in `workspace-full` and thus break Docker-in-Gitpod.

(Also, even if you _don't_ change any lines here, I believe that the base image change will be enough to nuke/rebuild all the Dazzle layers anyway.)

Thus, this Pull Request most likely depends on https://github.com/gitpod-io/gitpod/issues/3051 so either we first fix that issue, or we break Docker-in-Gitpod.

File: rethinkdb/Dockerfile
Comment (jankeromnes) [inline rethinkdb/Dockerfile:5]: RethinkDB is deprecated and I think this Dockerfile doesn't even get built anymore.

Please delete the entire `rethinkdb` directory.

**Review by jankeromnes — COMMENTED** [review]
Wow, awesome improvements here. Many thanks @aledbf! 💯

> The changes in the pull request also reduce the size of the images (between ~200MB and ~600MB)

That's super cool! Saving a few hundred MBs in these images massively helps with Gitpod ops (faster workspace start-up times, faster node initialization, less disk pressure, and various other benefits).

I wonder if some of the savings are due to building `workspace-base` without Dazzle. (I previously found that Dazzle seems to add a 330MB overhead to `workspace-full`: https://github.com/gitpod-io/workspace-images/pull/318#issuecomment-741812731)

A few minor suggestions in-line, but otherwise this seems good-to-go, provided we extensively test popular Gitpod projects with the new images, and in particular verify that Docker-in-Gitpod still works.

File: full/Dockerfile
Comment (jankeromnes) [reply full/Dockerfile:214]: Valid point from @csweichel on Slack:

> can’t we pin Docker do a known working version?

So if you pin the `docker-ce` and `docker-ce-cli` versions to what's currently live in Gitpod (it's `19.03.14` -- see https://github.com/gitpod-io/workspace-images/pull/331#issuecomment-768968566) that should solve the Docker problem (and remove the dependency on https://github.com/gitpod-io/gitpod/issues/3051)

File: rethinkdb/Dockerfile
Comment (aledbf) [reply rethinkdb/Dockerfile]: done

File: full/Dockerfile
Comment (aledbf) [reply full/Dockerfile]: Ubuntu package version is `3.16.3` and brew `3.19.5`

File: full/Dockerfile
Comment (aledbf) [reply full/Dockerfile]: fixed

File: full/Dockerfile
Comment (aledbf) [reply full/Dockerfile]: done

File: full/Dockerfile
Comment (aledbf) [reply full/Dockerfile]: fixed

**Comment by jankeromnes** [comment]
Thanks! CircleCI is 100% green, which means that all images have been published to Docker Hub with the tag `:branch-aledbf-base-image`, e.g.:

```
gitpod/workspace-full:branch-aledbf-base-image
```

Next we'll probably want to fork a few popular Gitpod repositories, then temporarily switch their image to the `:branch-aledbf-base-image` tag, and open them in Gitpod to verify that everything still works well.

**Comment by jankeromnes** [comment]
Note: This Pull Request upgrades Docker (CLI + daemon) from `19.03.14` to `19.03.15`. But I suspect that's okay:
- It's `20.10.2` that's known to break Docker-in-Gitpod: https://github.com/gitpod-io/gitpod/issues/3051
- I've double-checked that the [reported `docker-compose` breakage](https://github.com/gitpod-io/workspace-images/pull/331#issuecomment-768968566) doesn't reproduce with `19.03.15`

**Review by jankeromnes — APPROVED** [review]
Changes look good, bring a ton of value, and work (surprisingly) well! 🚀

Many thanks again 🙏

**Comment by aledbf** [comment]
@jankeromnes can we merge this change? (any change will break this PR)

**Comment by jankeromnes** [comment]
@aledbf yes please. 🚀

Note: We generally let authors merge their PRs once they're approved. Do you have write access?

(Also, no need to wait for the new CircleCI builds.)

File: base/install-packages
Comment (apolopena) [reply base/install-packages:20]: @jankeromnes Could this possibly be inadvertently causing this [issue](https://github.com/apolopena/gitpod-laravel8-starter/issues/46) I am having in all of my projects? Or [this line](https://github.com/gitpod-io/workspace-images/commit/6a8181b1c37f99b68aed2046abbe7c6381249187#r47850267) of code maybe?
I also posted the problem I am having with all my projects in the gitpod community forum [here](https://community.gitpod.io/t/var-log-apache2-not-found-by-default/2876). I don't mean to convolute or hijack this great PR. I just wanted to a stab at finding the root cause.
Thanks.

**Pull request merged**
